bigsi/cmds/large_build.py

```diff
@@ -0,0 +1,40 @@
+from bigsi.storage import get_storage
+from bigsi.graph.metadata import SampleMetadata
+from bigsi.bloom import BitMatrixGroupReader
+
+BLOOM_FILTERS_SIZE_KEY = "ksi:bloomfilter_size"
+NUM_HASH_FUNCTIONS_KEY = "ksi:num_hashes"
+NUM_ROWS_KEY = "number_of_rows"
+NUM_COLS_KEY = "number_of_cols"
+DB_INSERT_BATCH_SIZE = 1000
+
+
+def large_build(config, input_path_list, num_cols_list, sample_list):
+    storage = get_storage(config)
+    num_rows = int(config["m"])
+
+    with BitMatrixGroupReader(zip(input_path_list, num_cols_list), num_rows) as bmgr:
+        processed = 0
+        bit_arrays = []
+        keys = []
+        for row_index in range(num_rows):
+            keys.append(row_index)
+            bit_arrays.append(next(bmgr))
+            processed = processed + 1
+            if processed == DB_INSERT_BATCH_SIZE:
+                storage.set_bitarrays(keys, bit_arrays)
+                storage.sync()
```

Comment:
I am wondering if a single `storage.sync()` at the end of this loop would be more efficient than syncing every `DB_INSERT_BATCH_SIZE`. I think each DB deals differently with when it `sync`s with the disk (I think berkeleydb syncs when its cache is full), so I guess we should let it choose when it syncs or not? If its cache is never full, then it will `sync` a single time to the disk, while keeping `storage.sync()` here could make it `sync` several times. Anyway, these are just thoughts, which can be wrong as I don't know BIGSI much

Reply:
This is deliberate! What the `sync()` function does is to flush any cached items into the disk. The is deliberately called in order to keep the cache small and hence keep the memory usage low.

Not sure about whether different db implementations have mechanism to set maximum cached items, but I wanted to control over when this is done.